### PR TITLE
Support providing a Supplier for TaggedMetricRegistry

### PR DESCRIPTION
**Pull request by justinuang**
We want to use a SlidingTimeWindowArrayReservoir in our code base:
https://metrics.dropwizard.io/4.0.0/manual/core.html#sliding-time-window-reservoirs

### CODE REVIEW

**Comment by justinuang** [comment]
@iamdanfox can I get a review?

**Comment by iamdanfox** [comment]
Defering to @schlosna on this one - I don't actually have write on this repo ;)

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/TaggedMetricRegistry.java [hunk 1]
```diff
@@ -38,6 +39,8 @@
      */
     Timer timer(MetricName metricName);
 
+    Timer timer(MetricName metricName, Supplier<Timer> timerSupplier);
```
Comment (schlosna) [inline tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/TaggedMetricRegistry.java:42]: I don't know of any other implementations of `TaggedMetricRegistry` than `DefaultTaggedMetricRegistry` but we might want to implement these new methods via Java 8 `default`s here in interface for back compat

**Review by schlosna — COMMENTED** [review]
Can we add some tests using alternate suppliers (e.g. time window)?

File: tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/TaggedMetricRegistry.java (hunk 1 above)
Comment (schlosna) [reply tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/TaggedMetricRegistry.java:42]: I found one interface that extends `TaggedMetricRegistry` and one implementation of that extension, but I think we're ok to make this interface change without the `default` and just update that one usage.

Would still like to get test coverage over these new methods for hygiene, then happy to merge and tag a 0.10.0.

**Comment by justinuang** [comment]
I added some tests using mocks so we can assert the exact behavior we wanted (supplier being called only once). I wasn't too keen on typing the same test over and over so i pulled out a common test method. Let me know if you have any comments.

**Review by schlosna — APPROVED** [review]

**Comment by justinuang** [comment]
Awesome, can we get a release?

**Comment by schlosna** [comment]
@justinuang tagged 0.10.0 and should be released pending https://circleci.com/gh/palantir/tritium/258 build

**Pull request merged**